java/flight/src/test/java/org/apache/arrow/flight/TestLeak.java

```diff
@@ -0,0 +1,183 @@
+      while (!callFinished.get()) {
+        Thread.sleep(100);
+      }
+
+      s.shutdown();
+      s.awaitTermination();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Test
+  public void testCancelingDoPutDoesNotLeak() {
+    final AtomicBoolean callFinished = new AtomicBoolean();
+    try (final BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+        final FlightServer s =
+            FlightTestUtil.getStartedServer(
+                (location) -> FlightServer.builder(allocator, location, new LeakFlightProducer(allocator, callFinished))
+                    .build());
+        final FlightClient client = FlightClient.builder(allocator, s.getLocation()).build()) {
+
+      try (final VectorSchemaRoot root = VectorSchemaRoot.create(getSchema(), allocator)) {
+        final FlightDescriptor descriptor = FlightDescriptor.command(new byte[0]);
+        final SyncPutListener listener = new SyncPutListener();
+        final FlightClient.ClientStreamListener stream = client.startPut(descriptor, root, listener);
+        for (int batch = 0; batch < 1000; batch++) {
+          for (int col = 0; col < 11; col++) {
+            final Float8Vector vector = (Float8Vector) root.getVector(Integer.toString(col));
+            vector.allocateNew();
+            for (int row = 0; row < ROWS; row++) {
+              vector.setSafe(row, 10.);
+            }
+          }
+          root.setRowCount(ROWS);
+          // Unlike DoGet, this method fairly reliably will write the message to the stream, so even without the fix
+          // for ARROW-7343, this won't leak memory.
+          stream.putNext();
+        }
+        stream.completed();
+      }
+
+      s.shutdown();
+      s.awaitTermination();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A FlightProducer that always produces a fixed data stream with metadata on the side.
+   */
+  private static class LeakFlightProducer extends NoOpFlightProducer {
+
+    private final BufferAllocator allocator;
+    private final AtomicBoolean callFinished;
+
+    public LeakFlightProducer(BufferAllocator allocator, AtomicBoolean callFinished) {
+      this.allocator = allocator;
+      this.callFinished = callFinished;
+    }
+
+    @Override
+    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+      BufferAllocator childAllocator = allocator.newChildAllocator("foo", 0, Long.MAX_VALUE);
+      VectorSchemaRoot root = VectorSchemaRoot.create(TestLeak.getSchema(), childAllocator);
+      root.allocateNew();
+      listener.start(root);
+
+      // We can't poll listener#isCancelled since gRPC has two distinct "is cancelled" flags.
+      // TODO: should we continue leaking gRPC semantics? Can we even avoid this?
+      listener.setOnCancelHandler(() -> {
+        try {
+          for (int batch = 0; batch < 1000; batch++) {
```

Comment:
are these loops only essential for busy waiting?  Can you use Lock/Condition to make these more deterministic?